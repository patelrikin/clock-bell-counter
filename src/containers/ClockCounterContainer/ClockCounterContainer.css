.clock-container {
  position: relative;
}

.time-containers {
  display: flex;
  justify-content: center;
}

.time-containers label {
  padding-right: 10px;
}

.time-containers input {
  padding: 5px;
}

.start-time-container {
}

.end-time-container {
}

.error {
  color: red;
}