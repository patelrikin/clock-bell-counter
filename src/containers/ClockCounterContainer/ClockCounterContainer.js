import React, { Component } from 'react';
import { isValidHour, isValidMinute, calculateRingsBetweenTime } from '../../helpers/TimeHelpers';
import './ClockCounterContainer.css';

class ClockCounterContainer extends Component {
  constructor() {
    super();
    this.state = {
      bellCount: '',
      validSTART: true,
      validEND: true
    };
  }

  // Removes all non-digits character except ':' literal
  sanitizeUserInput = val => val.trim().replace(/[^\d:]/g, '');

  isValid = val => {
    let [result, hour, , minute] =
      this.sanitizeUserInput(val).match(/^(\d{1,2})([:]{1})(\d{2})$/) || []; // https://regex101.com/r/sxzzNg/2/
    return result
      ? isValidHour(hour) && isValidMinute(minute)
      : false;
  };

  updateInputErrorState = (targetName, validity) => this.setState({ [`valid${targetName.toUpperCase()}`]: validity });

  validateUserInput = ({ target } = {}) => {
    return this.updateInputErrorState(target.name, this.isValid(target.value));
  };

  updateDisplay = bellCount => this.setState({ bellCount });

  handleSubmit = e => {
    e.preventDefault();
    if (!(this.state.validSTART && this.state.validEND)) return;

    let startTime = document.querySelector("input[name='start']").value;
    let endTime = document.querySelector("input[name='end']").value;

    let result = calculateRingsBetweenTime(startTime, endTime);
    this.updateDisplay(result);
  };

  render() {
    return (
      <div className="clock-container p-20">
        <form onSubmit={this.handleSubmit}>
          <div className="time-containers p-20">
            <div className="start-time-container p-20">
              <label htmlFor="start">Start Time:</label>
              <input
                name="start"
                type="text"
                placeholder="HH:MM"
                required
                onBlur={this.validateUserInput}
              />
              {!this.state.validSTART && (
                <p className="error">Invalid Start Time.</p>
              )}
            </div>
            <div className="end-time-container p-20">
              <label htmlFor="end">End Time:</label>
              <input
                name="end"
                type="text"
                placeholder="HH:MM"
                required
                onBlur={this.validateUserInput}
              />
              {!this.state.validEND && (
                <p className="error">Invalid End Time.</p>
              )}
            </div>
          </div>
          <input type="submit" className="button"/>
        </form>
        <p>{this.state.bellCount}</p>
      </div>
    );
  }
}

export default ClockCounterContainer;
