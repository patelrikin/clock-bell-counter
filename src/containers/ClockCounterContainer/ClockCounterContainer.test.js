import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import ClockCounterContainer from './ClockCounterContainer';

let TestComponent;

describe('ClockCounterContainer', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<ClockCounterContainer />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('should have rendered using renderer()', () => {
    TestComponent = renderer.create(<ClockCounterContainer />);
    let tree = TestComponent.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
