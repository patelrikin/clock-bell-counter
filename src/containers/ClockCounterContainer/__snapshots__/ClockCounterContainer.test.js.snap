// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClockCounterContainer should have rendered using renderer() 1`] = `
<div
  className="clock-container p-20"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="time-containers p-20"
    >
      <div
        className="start-time-container p-20"
      >
        <label
          htmlFor="start"
        >
          Start Time:
        </label>
        <input
          name="start"
          onBlur={[Function]}
          placeholder="HH:MM"
          required={true}
          type="text"
        />
      </div>
      <div
        className="end-time-container p-20"
      >
        <label
          htmlFor="end"
        >
          End Time:
        </label>
        <input
          name="end"
          onBlur={[Function]}
          placeholder="HH:MM"
          required={true}
          type="text"
        />
      </div>
    </div>
    <input
      className="button"
      type="submit"
    />
  </form>
  <p>
    
  </p>
</div>
`;
